### PR TITLE
:lock::sparkles: Add time safe compare function

### PR DESCRIPTION
**Pull request by LionC**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -65,19 +65,30 @@ one of the three passed credentials.
...
 
 ```js
 app.use(basicAuth( { authorizer: myAuthorizer } ))
 
 function myAuthorizer(username, password) {
-    return username.startsWith('A') && password.startsWith('secret')
+    const userMatches = basicAuth.safeCompare(username, 'customuser')
+    const passwordMatches = basicAuth.safeCompare(password, 'custompassword')
+
+    return userMatches & passwordMatches
```
Comment (hraban) [inline README.md]: love it, very readable.

**Comment by LionC** [comment]
@hraban With this merged and released, would you mind updating / closing the (vulnerability on snyk)[https://snyk.io/vuln/SNYK-JS-EXPRESSBASICAUTH-174345]?

**Comment by feross** [comment]
Why does this PR change a bunch of instances of the boolean `&&` operator to the bitwise `&` operator? Is this necessary for correctness? Is this just an optimization for speed?

**Comment by LionC** [comment]
@feross It is necessary to guarantee that it runs in constant time to protect against timing attacks.

The `&&` operator shortcuts (it stops evaluating if the left side is `false`), while the `&` operator always evaluates both sides.

**Comment by feross** [comment]
That makes sense. A comment to that effect would have been nice. Thank you for explaining.

**Comment by dcousens** [comment]
`staticUsersAuthorizer` is still acts as a timing oracle for the number of `users`.

I don't know if `O(n)` would be preferred to the `O(n) worst case` as it is at the moment.
You could instead use `hash(user, 'base64')` to index into the array which should be relatively constant time and eliminate the need to do `safeCompare` on the username.

**Pull request merged**